src/sys/socket/mod.rs

```diff
@@ -287,18 +287,30 @@ impl fmt::Debug for Ipv6MembershipRequest {
     }
 }
 
-/// Copy the in-memory representation of src into the byte slice dst,
-/// updating the slice to point to the remainder of dst only. Unsafe
-/// because it exposes all bytes in src, which may be UB if some of them
-/// are uninitialized (including padding).
-unsafe fn copy_bytes<'a, 'b, T: ?Sized>(src: &T, dst: &'a mut &'b mut [u8]) {
+/// Copy the in-memory representation of `src` into the byte slice `dst`.
+///
+/// Returns the remainder of `dst`.
+///
+/// Panics when `dst` is too small for `src` (panics if `mem::size_of_val(src) >= dst.len()`).
+///
+/// Unsafe because it exposes all bytes in `src`, which may be UB if some of them are uninitialized
```

Comment:
It's not actually unsafe to read an uninitialized slice, because slices are always supposed to be initialized.  The unsafe part is creating the uninitialized slice in the first place.  However, it _is_ unsafe to call `copy_nonoverlapping` without verifying that the slices don't overlap, without verifying that `T` implements `Copy`, and without `drop`ing the contents of `dst`.  Please update the comment.

Reply:
The problem is that this is reading the bytes of an arbitrary `T`, including its padding. If reading the bytes from a `T` is safe, then so is this function. It doesn't create any uninitialized memory or slices.

AFAIK, LLVM considers reads of padding bytes `undef`, which would be unsafe, but I don't know if rustc makes it safe in some way (eg. by providing its own padding members that read as junk instead of `undef`).

Reply:
So the unsafe part is that it's effectively transmuting the `T`.  Transmutation is unsafe regardless of padding issues.

Reply:
I don't think transmutation to just bytes is unsafe except when padding is involved (as there's no way to read data that isn't valid as a `u8`), but I've clarified what's going on here